icu4c/source/common/uloc.cpp

```diff
@@ -1199,48 +1164,48 @@ ulocimp_getLanguage(const char *localeID,
 
     /* if it starts with i- or x- then copy that prefix */
     if(_isIDPrefix(localeID)) {
-        if(i<languageCapacity) {
-            language[i]=(char)uprv_tolower(*localeID);
-        }
-        if(i<languageCapacity) {
-            language[i+1]='-';
-        }
-        i+=2;
+        char c = (char)uprv_tolower(*localeID);
+        sink.Append(&c, 1);
+        sink.Append("-", 1);
         localeID+=2;
     }
 
     /* copy the language as far as possible and count its length */
     while(!_isTerminator(*localeID) && !_isIDSeparator(*localeID)) {
-        if(i<languageCapacity) {
-            language[i]=(char)uprv_tolower(*localeID);
-        }
-        if(i<3) {
-            U_ASSERT(i>=0);
-            lang[i]=(char)uprv_tolower(*localeID);
-        }
-        i++;
+        char c = (char)uprv_tolower(*localeID);
+        sink.Append(&c, 1);
         localeID++;
     }
 
-    if(i==3) {
+    if(sink.size()==3) {
```

Comment:
Let's try to collect the language subtag in an internal CharString and work with that until we have the final version and append that to the sink. Then we don't need the new methods.